### ISSUE TITLE
Failed jobs are attempting more than "attempts" value

### ISSUE THREAD

**Issue by wootwoot1234** [taskforcesh/bullmq#1571]
We are running BullMQ 3.2.4 and are seeing some failed jobs continue to retry more than the attempts value.

<img width="1598" alt="Screen Shot 2022-12-06 at 12 57 36 PM" src="https://user-images.githubusercontent.com/419609/205998662-89998e68-65aa-44fb-9eb7-8610aeb35bd4.png">

**Comment by tmhao2005** [comment]
In our code, we set a dynamic limiter as having exception:

```ts
worker.rateLimit(5_000);
throw Worker.RateLimitError();
```

then the job was moved to `waiting` list and the `attemptsMade` was increased greater than its max attempts value. 

Basically we can handle this sort of logic from our end but we think the expected result is the job getting moved to failed list instead.

**Comment by manast** [comment]
Not sure if that is the best thing to do in all cases. Since this is a special error that for example does not trigger backoffs, etc. The job has not really failed, it has just been moved back to wait as it is now considered rate-limited.

**Comment by manast** [comment]
I will think more about it...

**Comment by benmkramer** [comment]
I am having some similar issues. I see the job run both more than the defined attemps (seems to continuously retry no matter what... maybe that is expected since the rate limit is kind of a "try again later" more than a ... something went wrong and just want to try again)

Here is some code that can reproduce for me:
```javascript
const { Worker, Queue } = require("bullmq");
const Redis = require("ioredis");

const redis = new Redis("redis://localhost:6379", {
  maxRetriesPerRequest: null,
  enableReadyCheck: false,
});

const worker = new Worker("myQueue", async () => {
  console.log("running job...");
  const [isRateLimited, duration] = doSomething(); // always true, 1_000
  if (isRateLimited) {
    await worker.rateLimit(duration);
    throw Worker.RateLimitError();
  }
  console.log("done!"); // never reaches this point
}, {
  connection: redis,
  // WARNING: turning this on seems to run the job MANY MANY times
  // concurrency: 2, 
});

function doSomething() {
  return [true, 1_000];
}

const queue = new Queue("myQueue", { connection: redis });
queue.add("someJob", { foo: "bar" }, {
  attempts: 3,
});
```

Corresponding package.json dependencies:
```json
  {
    "bullmq": "^3.12.0",
    "ioredis": "^5.3.2"
  }
```

Another related problem (commented in code snippet above) is when using this rate limiting with a job that has concurrency defined, it seems to run some large number of times very quickly. It even seems to ignore any backoff configuration in that case as well, as if neither the `rateLimit` duration nor the backoff configuration are respected.

**Comment by manast** [comment]
> [edit] I think I missed a comment above, where it seems the backoff not being respected was already addressed. That's fair. However, curious if any guidance on handling concurrency for these jobs. Would still be great to run multiple jobs in parallel (via concurrency setting) and then pause them if hitting the rate limit.

What's the issue? AFAIK concurrency is handled correctly when manually rate limiting jobs.

**Comment by benmkramer** [comment]
> What's the issue? AFAIK concurrency is handled correctly when manually rate limiting jobs.

I put together a simple video. It's using the same script as above with some different console.logs, a longer rate limit duration, with settings of concurrency = 4, and attempts = 1. I first ran `redis-cli flushdb` to make sure the db didn't have some weird state of old jobs, then started this script which adds 1 item to the queue. It then seems to run indefinitely over and over.

https://user-images.githubusercontent.com/2603478/236506297-0fd12b7b-4c4b-4a38-95ed-415b02103662.mp4

**Comment by roggervalf** [comment]
hey @benmkramer in this case, you can check [attemptsMade](https://api.docs.bullmq.io/classes/Job.html#attemptsMade) property along with your isRateLimited variable, so you can decide when stopping to call worker.rateLimit method as this is a special case as mentioned by @manast, probably this could be a pattern, we will double check this case

**Comment by benmkramer** [comment]
Thanks for your reply, @roggervalf! I can see why attemptsMade would not necessarily be respected with rate limiting since it is a special case failure 👍 

I am still seeing unexpected behavior when using concurrency with manual rate limiting. Perhaps this should be a separate issue, happy to open if so. I packaged it into a small repo (https://github.com/benmkramer/bullmq-concurrency-manual-rate-limit) where the readme explains how I am set up and running this. Happy to connect and demonstrate if this is in fact some unusual behavior that should not happen and not just something I am doing wrong!

**Comment by WesleyYue** [comment]
Is there a way to differentiate between attempts due to rate limit errors and attempts due to other errors?

The behaviour I would expect (and I'm assuming the same for others in the thread) is:
1. attemptsMade only tracks failures due to errors that are not rate limits
2. the backoff and max attempt settings work as intended based on this attemptsMade var.

Separately, you can probably implement a feature to track rate limit attempts specifically, with a max rate limit attempt config, but this is not important to me.

**Comment by manast** [comment]
This issue should now be resolved in v5 https://bullmq.io/news/231221/bullmqv5-release/
